ztoc/compression/tar_zinfo.go

```diff
@@ -0,0 +1,198 @@
+	tarZinfoLength = 4 + 8 + 8
+	// `TarZinfo` version. consistent with `GzipZinfo` version
+	zinfoVersion = 2
+)
+
+// TarZinfo implements the `Zinfo` interface for uncompressed tar files.
+// It only needs a span size and tar file size, since a tar file is already
+// uncompressed.
+// For tar file, `compressed`-related concepts (e.g., `CompressedArchiveSize`)
+// are only to santisfy the `Zinfo` interface and equal to their `uncompressed`-equivalent.
+type TarZinfo struct {
+	version  int32
+	spanSize int64
+	size     int64
+}
+
+// newTarZinfo creates a new instance of `TarZinfo` from serialized bytes.
+func newTarZinfo(zinfoBytes []byte) (*TarZinfo, error) {
+	if len(zinfoBytes) != tarZinfoLength {
+		return nil, fmt.Errorf("invalid checkpoint length, expected: %d, actual: %d", tarZinfoLength, len(zinfoBytes))
+	}
+
+	var version int32
+	var spanSize, size int64
+
+	buf := bytes.NewReader(zinfoBytes)
+	if err := binary.Read(buf, binary.LittleEndian, &version); err != nil {
+		return nil, fmt.Errorf("invalid checkpoint data: %w", err)
+	}
+	if err := binary.Read(buf, binary.LittleEndian, &spanSize); err != nil {
+		return nil, fmt.Errorf("invalid checkpoint data: %w", err)
+	}
+	if err := binary.Read(buf, binary.LittleEndian, &size); err != nil {
+		return nil, fmt.Errorf("invalid checkpoint data: %w", err)
+	}
+
+	return &TarZinfo{
+		version:  version,
+		spanSize: spanSize,
+		size:     size,
+	}, nil
+}
+
+// newTarZinfoFromFile creates a new instance of `TarZinfo` given tar file name and span size.
+func newTarZinfoFromFile(tarFile string, spanSize int64) (*TarZinfo, error) {
+	fstat, err := os.Stat(tarFile)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get file stat: %w", err)
+	}
+
+	return &TarZinfo{
+		version:  zinfoVersion,
+		spanSize: spanSize,
+		size:     fstat.Size(),
+	}, nil
+}
+
+// Close doesn't do anything since there is nothing to close/release.
+func (i *TarZinfo) Close() {}
+
+// Bytes returns the byte slice containing the `TarZinfo`. Integers are serialized
+// to `LittleEndian` binaries.
+func (i *TarZinfo) Bytes() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	if err := binary.Write(buf, binary.LittleEndian, i.version); err != nil {
+		return nil, fmt.Errorf("failed to generate tar zinfo bytes: %w", err)
+	}
+	if err := binary.Write(buf, binary.LittleEndian, i.spanSize); err != nil {
+		return nil, fmt.Errorf("failed to generate tar zinfo bytes: %w", err)
+	}
+	if err := binary.Write(buf, binary.LittleEndian, i.size); err != nil {
+		return nil, fmt.Errorf("failed to generate tar zinfo bytes: %w", err)
+	}
+
+	zinfoBytes := buf.Bytes()
+	if len(zinfoBytes) != tarZinfoLength {
+		return nil, fmt.Errorf("invalid tar zinfo bytes length, expected: %d, actual: %d", len(zinfoBytes), tarZinfoLength)
+	}
+	return zinfoBytes, nil
+}
+
+// MaxSpanID returns the max span ID.
+func (i *TarZinfo) MaxSpanID() SpanID {
+	return SpanID(i.size / i.spanSize)
```

Comment:
Is this off by one if the size is a multiple of the spanSize?

Reply:
you mean if `i.size % i.spanSize == 0` then we should off by one right?

yeah I think you're right.